src/pages/library/datasetBuilder/CohortEditor.ts

```diff
@@ -0,0 +1,434 @@
+import { icon } from 'src/components/icons';
+import {
+  CriteriaType,
+  DatasetResponse,
+  DomainType,
+  ProgramDataListType,
+  ProgramDataRangeType,
+} from 'src/libs/ajax/DatasetBuilder';
+import colors from 'src/libs/colors';
+import * as Utils from 'src/libs/utils';
+import { PAGE_PADDING_HEIGHT, PAGE_PADDING_WIDTH } from 'src/pages/library/datasetBuilder/constants';
+import {
+  AnyCriteria,
+  Cohort,
+  CriteriaGroup,
+  DatasetBuilderState,
+  DomainCriteria,
+  newCriteriaGroup,
+  ProgramDataListCriteria,
+  ProgramDataRangeCriteria,
+} from 'src/pages/library/datasetBuilder/dataset-builder-types';
+import { HomepageState, OnStateChangeType } from 'src/pages/library/datasetBuilder/DatasetBuilder';
+import { datasetBuilderCohorts } from 'src/pages/library/datasetBuilder/state';
+
+type CriteriaViewProps = { criteria: AnyCriteria; deleteCriteria: (criteria: AnyCriteria) => void };
+
+const CriteriaView = ({ criteria, deleteCriteria }: CriteriaViewProps) => {
+  return div(
+    {
+      style: {
+        display: 'flex',
+        width: '100%',
+        justifyContent: 'space-between',
+        alignItems: 'baseline',
+        paddingBottom: 5,
+        marginTop: 10,
+        marginBottom: 5,
+        borderBottom: `1px solid ${colors.dark(0.35)}`,
+      },
+    },
+    [
+      div({ style: { margin: '5px', display: 'flex', alignItems: 'center' } }, [
+        h(
+          Link,
+          {
+            'aria-label': 'delete criteria',
+            onClick: () => {
+              deleteCriteria(criteria);
+            },
+          },
+          [icon('minus-circle', { size: 24, style: { color: colors.danger() } })]
+        ),
+        div({ style: { marginLeft: 5 } }, [
+          Utils.cond(
+            [
+              'domainType' in criteria,
+              () => {
+                const domainCriteria = criteria as DomainCriteria;
+                return h(Fragment, [strong([`${domainCriteria.domainType.category}:`]), ` ${domainCriteria.name}`]);
+              },
+            ],
+            [
+              'listType' in criteria,
+              () => {
+                const listCriteria = criteria as ProgramDataListCriteria;
+                return h(Fragment, [strong([`${criteria.name}:`]), ` ${listCriteria.value.name}`]);
+              },
+            ],
+            [
+              'rangeType' in criteria,
+              () => {
+                const rangeCriteria = criteria as ProgramDataRangeCriteria;
+                return h(Fragment, [strong([`${criteria.name}:`]), ` ${rangeCriteria.low} - ${rangeCriteria.high}`]);
+              },
+            ],
+            [Utils.DEFAULT, () => div(['Unknown criteria type'])]
+          ),
+        ]),
+      ]),
+      `Count: ${criteria.count}`,
+    ]
+  );
+};
+
+export const renderCriteriaView = (deleteCriteria: (criteria: AnyCriteria) => void) => (criteria: AnyCriteria) =>
```

Comment:
I have a strong desire to move us away from renderXxxxx helper functions and define things as standard react functional components.  Looking at this specific case, this could be renamed KeyedCriteraView: React.FC<KeyedCriteriaViewProps>, but seeing as how it only apears to be used once inside an _.map loop, I would propose removing this helper fn entirely that only outputs a one-liner.... just do this inline in the _.map.

Reply:
This is a helper method (that maybe has a misleading name) that curries up the `deleteCriteria` parameter separately from the `criteria` parameter. This is nice because `deleteCriteria` is the same for all `CriteriaView`s and a curried function is a good way to express that.

Using a curried function allows for less code in the `map` as well, since you no longer have to turn a two-arg function into a one-arg function. Before:

```
_.map(
  renderCriteriaView((criteria: AnyCriteria) =>
    updateCohort(_.set(`criteriaGroups.${index}.criteria`, _.without([criteria], criteriaGroup.criteria)))
  ),
criteriaGroup.criteria)
```

after:
```
 _.map(
  (criteria: AnyCriteria) =>
    h(CriteriaView, {
      deleteCriteria: (criteria: AnyCriteria) =>
        updateCohort(
          _.set(`criteriaGroups.${index}.criteria`, _.without([criteria], criteriaGroup.criteria))
        ),
        criteria,
        key: criteria.id,
      }),
criteriaGroup.criteria)
```

IMO the curried form is easier to write and read, but I can change it if you're opposed to it.

Reply:
The use of currying continues to be a controversial topic.  Folks who are practiced in it's use and comfortable with it, do find it's capabilities powerful in allowing for brevity.  But for those who are not comfortable, the currying does add friction to reading and understanding what code is doing, especially since it is not immediately obvious that a given function is curried.  This is a pretty consistent friction point for onboarding as well.  And even the original author might struggle to remember the specific currying bits after being away from the code for a few months.

I would favor a version of your 'after' snippet that defines onDeleteCriteria handler before the render block, and then this code snippet can be a lot shorter:

```
 _.map(
  (criteria: AnyCriteria) =>
    h(CriteriaView, { deleteCriteria: onDeleteCriteria, criteria, key: criteria.id }),
  criteriaGroup.criteria)
```
  
Yes, curried functions, and fp have legitimate advantages.  But in general, I believe that curried functions are often not worth the squeeze, and can often be an over-optimization where the goal should be to favor simpler and more approachable code patterns.  Pile on the added effort to properly Typescript the curried functions, and... yeah.

We have not gone beyond "leave it up to teams to decide if they want to use currying/fp patterns or favor non-fp".  So if your team is in the comfortable-with-fp camp. then I guess keep the renderXxxx method if you prefer to delay considering the arguments above... But at a minimum, methods like that should be defined OUTSIDE the React.FC's function body, since that's the render loop and should be mindful of performance.

Reply:
For now I'm leaving it in the curried form. As far as I can tell the code is equivalent and I find the curried way is easier to read. If it comes up again it's not hard to change it to inline the operation instead.